#!/bin/bash
# meryl_genomescope 0.0.1
# Generated by dx-app-wizard.

set -x -e -o pipefail


main() {

    sudo chmod 777 /usr/bin/meryl

    echo "Value of fastq: '${fastq[@]}'"
	echo "Value of kmer: '$kmer'"
	echo "Value of read_length: '$read_length'"
	echo "Value of kmer_max: '$kmer_max'"

    for i in ${!fastq[@]}
    do
        
		one_fastq_jobs+=($(dx-jobutil-new-job -ikmer="$kmer" -ifastq="${fastq[$i]}" meryl_fastq))
        
    done
 
	printf -v ids_d ' %s:fastq_meryl' "${one_fastq_jobs[@]}"
	ids_d=${ids_d:1}
    
    echo $ids_d

    for one_fastq_jobs in "${one_fastq_jobs[@]}"; do 
        merge_fastq_args+=(-icount_kmer="$one_fastq_jobs":fastq_meryl)
    done

    merge_job=$(dx-jobutil-new-job "${merge_fastq_args[@]}" -ikmer="$kmer" -iread_length="$read_length" -ikmer_max="$kmer_max" union_meryl)
    dx-jobutil-add-output output_genomescope "$merge_job":output_genomescope --class=array:jobref
    
}

union_meryl(){

    sudo chmod 777 /usr/bin/meryl
    mkdir all_count
    cd all_count

    for i in $(dx ls); do 
        ending=$(echo "$i" | grep '/$' || echo ""); 
        if [ -n "$ending" ]; then 
            meryl_count_kmer+=($ending)
            dx download -r "$ending" 
        fi 
    done

                
    ulimit -Sn 32000
    meryl union-sum output union_meryl "${meryl_count_kmer[@]}"
   
    meryl histogram union_meryl | sed 's/\t/ /g' > union_meryl.hist
    
    echo union_meryl.hist
    
    Rscript /scripts/genomescope.R union_meryl.hist $kmer $read_length union_meryl_gs $kmer_max 

    mkdir -p ~/out/output_genomescope
    echo "test" > ~/out/output_genomescope/test.txt
    mv union_meryl_gs ~/out/output_genomescope/

    dx-upload-all-outputs --parallel

}

meryl_fastq() {
	
	sudo chmod 777 /usr/bin/meryl

	mem_in_gb=`head -n1 /proc/meminfo | awk '{print int($2*0.8/1024/1024)}'`
	echo $mem_in_gb
	
    dx-download-all-inputs

	mkdir -p ~/out/fastq_meryl/
	
	meryl count k="$kmer" output ~/out/fastq_meryl/${fastq_name%.fastq.gz}.meryl "${fastq_path}" memory=$mem_in_gb

    dx-upload-all-outputs --parallel
}