#!/bin/bash
# meryl_genomescope 0.0.1
# Generated by dx-app-wizard.

set -x -e -o pipefail

# Basic execution pattern: Your app will run on a single machine from
# beginning to end.
#
# Your job's input variables (if any) will be loaded as environment
# variables before this script runs.  Any array inputs will be loaded
# as bash arrays.
#
# Any code outside of main() (or any entry point you may add) is
# ALWAYS executed, followed by running the entry point itself.
#
# See https://wiki.dnanexus.com/Developer-Portal for tutorials on how
# to modify this file.

main() {

    sudo chmod 777 /usr/bin/meryl

    echo "Value of fastq: '${fastq[@]}'"
	echo "Memory: $(mem)"

    for i in ${!fastq[@]}
    do
        
		one_fastq_jobs+=($(dx-jobutil-new-job --instance-type="mem2_hdd2_x4" -ikmer="$kmer" -ifastq="${fastq[$i]}" meryl_fastq))
        
    done
 
	printf -v ids_d ' %s:fastq_meryl' "${one_fastq_jobs[@]}"
	ids_d=${ids_d:1}
    
    echo $ids_d
    
    #ulimit -Sn 32000
    #meryl union-sum output union_meryl $ids_d
   
    #meryl histogram union_meryl | sed 's/\t/ /g' > union_meryl.hist
    
    #echo union_meryl.hist
    
    #Rscript /resources/scripts/genomescope.R union_meryl.hist $kmer $read_length union_meryl_gs $kmer_max 

	mkdir -p ~/out/output_genomescope
	echo "test" > ~/out/output_genomescope/test.txt
	#mv union_meryl_gs ~/out/output_genomescope/

	dx-upload-all-outputs --parallel

}

meryl_fastq() {
	
	sudo chmod 777 /usr/bin/meryl
	
    dx-download-all-inputs

	mkdir -p ~/out/fastq_meryl/
	
	meryl count k="$kmer" output ~/out/fastq_meryl/${fastq_name%.fastq.gz}.meryl "${fastq_path}" memory=5

    dx-upload-all-outputs --parallel
}