<!-- dx-header -->

# 10X Longranger Align (DNAnexus Platform App)

The longranger align pipeline performs all of the functions of the longranger basic pipeline, plus aligns the reads with BWA and infers original input molecule extents. For more details, see the ALIGN Pipeline Map.
To run longranger align, the following parameters are required:

| Argument | Longranger Command  |  Description |
| --- | --- | --- |
| Output Prefix |--id  |  A unique run ID string: e.g. sample345 |
| FASTQ Input Files |--fastqs  |  Path of the FASTQ folder generated by longranger mkfastq |
e.g. /home/jdoe/runs/HAWT7ADXX/outs/fastq_path
| Sample Name |--sample  |  (optional) Sample name as specified in the sample sheet supplied to mkfastq. |
| Output format |--bam or none  | (optional) Output as unaligned BAM instead of FASTQ. |

The output of the pipeline will contain the main pipeline output files:

|File Name   |Description|
| --- | --- |
|summary.csv |Run summary metrics in CSV format|
|possorted_bam.bam   |Aligned reads annotated with barcode information|
|possorted_bam.bam.bai    | Index for possorted_bam.bam|
